File: part1.py
# -*- coding: utf-8 -*-
"""Project, part 1"""
import numpy as np
import matplotlib.pyplot as plt
#import rw #Assumes rwmodule has been compiled with f2py to produce rw.so
from rw import rwmodule as rw


def analyze_rnet(Ntime,Nm,X0,N0,L,Nt,display):
	"""Input variables:
	Ntime: number of time steps
    	Nm: number of walks
    	X0: initial node, (node with maximum degree if X0=0)
    	N0,L,Nt: recursive network parameters
    	"""
    	#Call rwnet
    	X=rw.rwnet(Ntime,Nm,X0,N0,L,Nt,1)
    	X0=X[1,1]
    	#Create an empty F function and K array for maximum values
    	F=np.zeros(((Ntime+1),(N0+Nt)))
    	K=np.zeros(Ntime+1)

        #Loop over the time and then steps to fill in thr F(t,n)
    	for i in range(Ntime+1):
    	    for j in range(Nm):
    	        #Increasing the appropriate elements of F(t,n)
    	        a=X[i,j]-1
    	        F[i,a]=F[i,a]+1
            #Making F a probability matrix
    	    F[i,:]=F[i,:]/(1.0*Nm)
    	
    	#Loop over all times to get the greatest number of walkers at each step    
        for k in range(Ntime+1):
            K[k] = np.argmax(F[k,:])
            
        #Assiging x axis for the plot
    	x=np.arange(1,(Ntime+2))

        #If check for the display flag
    	if (display==True):
    	    #Plotting the figure
    	    plt.figure()
    	    plt.plot(x, K)
    	    #Legend of the plot
            plt.legend(['Maximum amount of walkers'])
            plt.xlabel('x range')
            plt.ylabel('F(t,n)')
            plt.title('Evgeniia Gleizer. Created by analyze_rnet.')
    	    plt.show()
        return F



def convergence_rnet(Ntime,Nm,X0,N0,L,Nt,display):
    	"""Input variables:
	Ntime: number of time steps
    	m: number of walks
    	X0: initial node, (node with maximum degree if X0=0)
    	N0,L,Nt: recursive network parameters
    	"""
    	#I am presenting here a graph of the average fraction of the walkers 
    	#Only of the nodes which fraction of the walkers is non-zero
    	#against m. It is easy to see that with Ntime=105 and m increasing,
    	#it tendes to converge to around 0
    	
    	#M=np.array([10,40,50,60,100])
    	l=len(Nm)
    	x=np.zeros(l)
    	y=np.zeros(l)
    	for j in range(l):
    	    F=analyze_rnet(105,Nm[j],0,5,2,200, False)
    	    #rw.rwnet(Ntime,Nm,X0,N0,L,Nt,1)
    	    #print sum(F[:,(X0-1)])
    	    for i in range (Ntime+1):
    	        if (F[i,(X0-1)]>0):
    	            x[j]=x[j]+1
    	    y[j]=sum(F[:,(X0-1)])/x[j]  
    	    
    	if (display==True):
    	    plt.figure()
    	    plt.ylim(0,0.2)
    	    plt.plot(Nm, y)
    	    #Legend of the plot
            plt.legend(['Average amount of walkers against Nm'])
            plt.xlabel('Nm')
            plt.ylabel('Av. walkers')
            plt.title('Evgeniia Gleizer. Created by convergence_rnet.')
    	    plt.show()  	
#This function helps us see that the maxes of F tend to converge to a
#relatively small number  	
    	
if __name__== '__main__':
#add code here to call functions and generate figures
     #Ntime,Nm,X0,N0,L,Nt,display)
    Ntime = 1e5
    Nm=1e3
    N0= 5
    L=2
    Nt=200
    analyze_rnet(Ntime,Nm,0,N0,L,Nt,True)
    convergence_rnet(Ntime,[10,20,30,40,100,1000],0,N0,L,Nt,True)

File: part1.f90
!Project part 1
module rwmodule
    use network

contains



subroutine rwnet(Ntime,Nm,X0,N0,L,Nt,isample,X)
    !random walks on a recursive network
    !Input variables:
    !Ntime: number of time steps
    !m: number of walks
    !X0: initial node, (node with maximum degree if X0=0)
    !N0,L,Nt: recursive network parameters
    !Output: X, m random walks on network each with Ntime steps from initial node, X0
    !XM: fraction of walks at initial node, computed every isample timesteps
	implicit none
    integer :: i, i1, i2, i3, i4, i5, i6, i7, node, y, y1, y2, k,q,n,z, o
    integer, intent(in) :: Ntime,Nm,N0,L,Nt,X0,isample
    real(kind=8) :: u, cp
    real(kind=8), allocatable, dimension(:) :: prob
    integer, dimension(Ntime+1,Nm), intent(out) :: X
    integer :: qmax
    integer, dimension(N0+Nt) :: qnet, a
    integer, dimension(1) :: maxl
    integer, dimension(N0+L*Nt,2) :: enet 
    integer, dimension(size(enet,1)*2) :: alist1
    integer, dimension(size(qnet)) :: alist2
    integer, dimension(N0+Nt,N0+Nt) :: anet
    real(kind=8), dimension((Ntime+1)/isample) :: XM
    !Calling all the functions we will later use	
       call generate(N0,L,Nt,qmax,qnet,enet)
       call adjacency_list(qnet,enet,alist1,alist2)
       call adjacency_matrix(N0+Nt,enet,anet)
       allocate(prob(size(qnet)))
    
    !Loop for all the walks   
    do i=1,Nm
       X(1,i)=X0
       
       q=0
       !Loop to assign X0 to the maximum node in case it is given as 0
       if (X0==0) then 
          maxl=maxloc(qnet)
          X(1,i)=maxl(1)
       end if 

       !Loop over all steps
       do i3=2,(Ntime+1)
                    prob=0
                    !Finding the node we going to work with
                    k=X((i3-1),i)
                    !Assigning probability(to array, containing zeros)
                    prob=dble(anet(k,:))/qnet(X((i3-1),i))
                        
	            u=0
	            !Calling the random number to find probability
	            call random_number(u)
	            cp=0
                    !Loop to go over cumulitative probability and find the node for the next step
	            do i4=1,size(prob)
	               cp=cp+prob(i4)
	                if (u<=cp) exit
                    end do 
                !Assigning found node to the next element in the matrix    
                X(i3,i)=i4                   	 
	end do 
    end do 
        deallocate(prob)
        

   
     XM=0
     !Loop over all Ntimes to find Xm
     do i6=1,(Ntime+1),isample
         !Second part of the loop - over all Nm's
         do i7=1,Nm
             !Condition for adding the node to the ones we are interested in 
             if (X(i6,i7)==X0) then
              XM(i6)=XM(i6)+1  
             end if
         end do
     end do 
     XM=XM/Nm
     
    print *, X

end subroutine rwnet


subroutine rwnet_omp(Ntime,Nm,X0,N0,L,Nt,isample,numthreads,X)
    !parallelized version of rwnet, parallel regions should
    !use numthreads threads
    	implicit none
    integer :: i, i1, i2, i3, i4, i5, i6, i7, node, y, y1, y2, k,q,n,z, o
    integer, intent(in) :: Ntime,Nm,N0,L,Nt,X0,isample, numthreads
    real(kind=8) :: u, cp
    real(kind=8), allocatable, dimension(:) :: prob
    integer, dimension(Ntime+1,Nm), intent(out) :: X
    integer :: qmax
    integer, dimension(N0+Nt) :: qnet, a
    integer, dimension(1) :: maxl
    integer, dimension(N0+L*Nt,2) :: enet 
    integer, dimension(size(enet,1)*2) :: alist1
    integer, dimension(size(qnet)) :: alist2
    integer, dimension(N0+Nt,N0+Nt) :: anet
    real(kind=8), dimension((Ntime+1)/isample) :: XM
	
    call generate(N0,L,Nt,qmax,qnet,enet)
    call adjacency_list(qnet,enet,alist1,alist2)
    call adjacency_matrix(N0+Nt,enet,anet)
    allocate(prob(size(qnet)))
    ! Call the omp with the number of threads
    !$ call omp_set_num_threads(numThreads)
    
    !Parallising the loop
    !$OMP parallel do private(i,i3,prob,k,u,cp)
    !Loop for all the walks   
    do i=1,Nm
       X(1,i)=X0
       
       q=0
       !Loop to assign X0 to the maximum node in case it is given as 0
       if (X0==0) then 
          maxl=maxloc(qnet)
          X(1,i)=maxl(1)
       end if 

       !Loop over all steps
       do i3=2,(Ntime+1)
                    prob=0
                    !Finding the node we going to work with
                    k=X((i3-1),i)
                    !Assigning probability(to array, containing zeros)
                    prob=dble(anet(k,:))/qnet(X((i3-1),i))
                        
	            u=0
	            !Calling the random number to find probability
	            call random_number(u)
	            cp=0
                    !Loop to go over cumulitative probability and find the node for the next step
	            do i4=1,size(prob)
	               cp=cp+prob(i4)
	                if (u<=cp) exit
                    end do 
                !Assigning found node to the next element in the matrix    
                X(i3,i)=i4                   	 
	end do 
    end do 
    
        !Finishing parallelization
        !$OMP end parallel do 
        deallocate(prob)
   

 !Ntime+1,Nm  
     XM=0
     
     !Parallelizing the Xm loop
    !$OMP parallel do private(i6,i7)
     do i6=1,(Ntime+1),isample
         do i7=1,Nm
             if (X(i6,i7)==X0) then
              XM(i6)=XM(i6)+1  
             end if
         end do
     end do
     !Finishing parallelization
     !$OMP end parallel do  
     XM=XM/Nm
     


end subroutine rwnet_omp


end module rwmodule
